lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,79 +3642,1099 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+  SILType getRemappedTangentType(SILType type) {
+    return SILType::getPrimitiveType(
+        getTangentSpace(remapType(type).getASTType())->getCanonicalType(),
+        type.getCategory());
+  }
+
+  //--------------------------------------------------------------------------//
+  // Managed tangent value mapping
+  //--------------------------------------------------------------------------//
+
+  AdjointValue makeZeroTangentValue(SILType type) {
+    return AdjointValue::createZero(allocator, remapType(type));
+  }
+
+  /// Initializes an original value's corresponding tangent value. Its tangent
+  /// value must not be present before this function is called.
+  void initializeTangentValue(SILBasicBlock *origBB, SILValue originalValue,
+  AdjointValue tangentValue) {
+    assert(origBB->getParent() == original);
+    assert(originalValue->getType().isObject());
+    assert(tangentValue.getType().isObject());
+    assert(originalValue->getFunction() == original);
+    // The tangent value must be in the tangent space.
+    assert(tangentValue.getType() ==
+           getRemappedTangentType(originalValue->getType()));
+    auto insertion =
+        tangentValueMap.try_emplace(originalValue, tangentValue);
+    assert(insertion.second && "Tangent value inserted before");
+  }
+
+  /// Get the tangent for an original value. The given value must be in the
+  /// original function.
+  ///
+  /// This method first tries to find an entry in `tangentValueMap`. If a tangent
+  /// doesn't exist, create a zero tangent.
+  AdjointValue getTangentValue(SILValue originalValue) {
+    assert(originalValue->getType().isObject());
+    assert(originalValue->getFunction() == original);
+    auto insertion = tangentValueMap.try_emplace(
+        originalValue, makeZeroTangentValue(
+        getRemappedTangentType(originalValue->getType())));
+    auto it = insertion.first;
+    return it->getSecond();
+  }
+
+  /// Add a tangent value for the given original value.
+  void addTangentValue(SILBasicBlock *origBB, SILValue originalValue,
+                       AdjointValue newTangentValue) {
+    assert(originalValue->getType().isObject());
+    assert(newTangentValue.getType().isObject());
+    assert(originalValue->getFunction() == original);
+    LLVM_DEBUG(getADDebugStream() << "Adding tangent for " << originalValue);
+    // The tangent value must be in the tangent space.
+    assert(newTangentValue.getType() ==
+           getRemappedTangentType(originalValue->getType()));
+    auto insertion =
+        tangentValueMap.try_emplace(originalValue, newTangentValue);
+    auto inserted = insertion.second;
+    assert(inserted && "The Tangent value should not already exist.");
+    return;
+  }
+
+  //--------------------------------------------------------------------------//
+  // Differential generation helpers
+  //--------------------------------------------------------------------------//
+
+  void visitReturnInstDifferential(ReturnInst *ri) {
```

Comment:
I'd suggest renaming all of these differential emission methods to `emitTangentForXXXInst`.

Reply:
Should I keep the section comment above as `Differential generation helpers` or could `Tangent generation helpers` be better?

Reply:
`Tangent emission` would be good.